core/vm/vmcontext/caller.go

```diff
@@ -0,0 +1,84 @@
+package vmcontext
+
+import (
+	"math/big"
+
+	"github.com/celo-org/celo-blockchain/common"
+	"github.com/celo-org/celo-blockchain/core/types"
+	"github.com/celo-org/celo-blockchain/core/vm"
+	"github.com/celo-org/celo-blockchain/params"
+)
+
+type evmCaller struct {
+	vmConfig    vm.Config
+	vmContext   vm.Context
+	chainConfig *params.ChainConfig
+	state       vm.StateDB
+
+	dontMeterGas bool
+}
+
+func NewEVMCallerForCurrentBlock(chain ExtendedChainContext) (vm.EVMCaller, error) {
+	header := chain.CurrentHeader()
+	// FIXME small race condition here (Head changes between get header & get state)
```

Comment:
Can we fix this?

Reply:
I tried. problem is that `LightChain` also implements `ExtendedChainContext` and it doesn't have `StateAt()`. Maybe i can add a new method called `CurrentHeaderAndState()` to both

Reply:
I think adding this extra method this is the only way to be safe, you have to get both at once.

Reply:
Or we could expose the blockchain lock, but I don't think that is a sensible option.

Reply:
yeah, i would wait for now, since if you take a look at the next PR the Blockchain and Lightchain objects are the one implementing the Factory interface and this factory is either not used, or only used on a single ocassion, so we can revisit there.

Either removing this altogheter, or adding the getStateAndHeader option to blockchain and lightchain